File: src/main/java/PythonSol/LinearDataStructures/Lists/DoublyLinkedList/DoublyLinkedList.py
from typing import Any


class Node:
    def __init__(self, value: Any, prev: 'Node' = None, next: 'Node' = None) -> None:
        self.value: Any = value
        self.prev: 'Node' = prev
        self.next: 'Node' = next

class DoublyLinkedList():
    def __init__(self) -> None:
        self.size:int = 0

        self.head: Node = None
        self.last: Node = None

    def __len__(self) -> int:
        return self.size

    def append(self, value: Any) -> None:
        if self.head == None and self.last == None:
            self.head = self.last = Node(value)
        else:
            if self.size == 1:
                self.last = Node(value)

                self.last.prev = self.head
                self.head.next = self.last
            else:
                new_node: Node = Node(value)

                new_node.prev = self.last

                self.last.next = new_node
                self.last = new_node

        self.size += 1

    def delete(self) -> None:
        if self.head == None:
            raise Exception("The doubly linked list is empty")
        elif self.size == 1:
            self.head = self.last = None
        elif self.size == 2:
            self.head.next = None
            self.last = self.head
        else:
            prev: Node = self.last.prev
            prev.next = None
            self.last = prev

        self.size -= 1

    def delete_index(self, index: int) -> None:
        if index < 0 or index >= self.size:
            raise IndexError("The given index {0} is out of bounds or the doubly linked list is empty (size: {1})".format(index, self.size))

        if index == 0:
            if self.size == 1:
                self.head = self.last = None
            elif self.size == 2:
                self.last.prev = None
                self.head = self.last
            else:
                self.head.next.prev = None
                self.head = self.head.next
        elif index == self.size - 1:
            if self.size == 2:
                self.head.next = None
                self.last = self.head
            else:
                self.last.prev.next = None
                self.last = self.last.prev
        else:
            counter:int = 0
            previous_node: Node = None
            current_node: Node = self.head

            while counter != index:
                previous_node = current_node
                current_node = current_node.next

                counter += 1

            current_node.next.prev = previous_node
            previous_node.next = current_node.next

        self.size -= 1

    def search(self, value: Any) -> Node:
        if self.head == None:
            raise Exception("The doubly linked list is empty.")

        current: Node = self.head
        while current != None:
            if current.value == value:
                return current

            current = current.next 

        raise Exception("No node found in the doubly linked list that has the given value : {0}".format(value))

    def print_list(self) -> None:
        if self.head == None:
            print("The doubly linked list is empty.")
            return

        current: Node = self.head
        while current != None:
            print(current.value, end = " ")
            current = current.next

        print()



if __name__ == "__main__":
    dll = DoublyLinkedList()
    
    for i in range(1, 6):
        dll.append(i)

    for i in range(1, 6):
        print(dll.search(i).value)

File: src/main/java/CCPPSol/LinearDataStructures/Lists/DoublyLinkedList/DoublyLinkedList.cpp
#include <iostream>

using namespace std;


class Node {
public:
    int value;
    Node* next;
    Node* prev;

    Node(int value) {
        this->value = value;
    }

    Node(int value, Node* next) {
        this->value = value;
        this->next = next;
    }

    Node(int value, Node* next, Node* prev) {
        this->value = value;
        this->next = next;
        this->prev = prev;
    }
};


class DoublyLinkedList {
private:
    int size = 0;
public:
    Node* head;
    Node* last;

    int getSize() {
        return this->size;
    }

    void append(int value) {
        if(this->head == nullptr) {
            this->head = this->last = new Node(value);
        } else if (this->size == 1) {
            this->last = new Node(value);
            
            this->last->prev = this->head;
            this->head->next = this->last;
        } else {
            Node* new_node = new Node(value);

            new_node->prev = this->last;
            this->last->next = new_node;
            this->last = new_node;
        }

        this->size += 1;
    }

    void delete_last() {
        if(this->size == 1) {
            this->head = nullptr;
            this->last = nullptr;
        } else if (this->size == 2) {
            this->head->next = nullptr;
            this->last = this->head;
        } else {
            Node* prev = this->last->prev;
            prev->next = nullptr;
            this->last = prev;
        }

        this->size -= 1;
    }

    void delete_index(int index) {
        if (index == 0) {
            if (this->size == 1) {
                this->head = nullptr;
                this->last = nullptr;
            } else if (this->size == 2) {
                this->last->prev = nullptr;
                this->head = this->last;
            } else {
                this->head->next->prev = nullptr;
                this->head = this->head->next;
            }
        } else if (index == this->size - 1) {
            if (this->size == 2) {
                this->head->next = nullptr;
                this->last = this->head;
            } else {
                this->last->prev->next = nullptr;
                this->last = this->last->prev;
            }
        } else {
            int counter = 0;
            Node* previous_node = nullptr;
            Node* current_node = this->head;

            while(counter != index) {
                previous_node = current_node;
                current_node = current_node->next;
                counter += 1;
            }

            current_node->next->prev = previous_node;
            previous_node->next = current_node->next;
        }

        this->size -= 1;
    }

    Node* search(int value) {
        if (this->head == nullptr) {
            cout << "The doubly linked list is empty." << endl;
            return NULL;
        }

        Node* current = this->head;
        while (current != NULL) {
            if (current->value == value) {
                return current;
            }

            current = current->next;
        }

        cout << "No node with the value " << value << " could be found in the doubly linked list" << endl;
        return NULL;
    }

    void print_list() {
        if(this->head == nullptr) {
            cout << "The linked list is empty." << endl;
            return;
        }

        Node* current = this->head;
        while (current != NULL) {
            cout << current->value << " ";

            current = current->next;
        }

        cout << " " << endl;
    }
};


int main() {
    DoublyLinkedList* dll = new DoublyLinkedList();

    for(int i = 1; i <= 5; ++i) {
        dll->append(i);
    }

    for(int i = 1; i <= 5; ++i) {
        cout << dll->search(i)->value << endl;
    }

    return 0;
}